File: dart/rosetta_code/Read_a_configuration_file/Read_a_configuration_file.dart
/**
 * Solution for Rosetta Code task ["Read a configuration file"]
 * (http://rosettacode.org/wiki/Read_a_configuration_file)
 */

#import('dart:io');


main() {
  var file = new File('fruit.cfg');
  var fullname = '';
  var favouritefruit = '';
  var needspeeling = false;
  var seedsremoved = false;
  var otherfamily = [];

  file.readAsLines().then((List<String> lines) {
    final blankOnly = const RegExp(@'^(#|;|\s*$)');
    final keyValue = const RegExp(@'\s*([^ =]+)([ =].*)?$');

    for (var line in lines) {
      if (!line.contains(blankOnly)) {
        final match = keyValue.firstMatch(line);
        final key = match[1].toLowerCase();
        final value = match[2];

        switch (key) {
          case 'fullname': fullname = value.trim(); break;
          case 'favouritefruit': favouritefruit = value.trim(); break;
          case 'needspeeling': needspeeling = true; break;
          case 'seedsremoved': seedsremoved = true; break;
          case 'otherfamily':
            otherfamily = value.split(',').map((s) => s.trim());
            break;
        }
      }
    }

    print('fullname = $fullname');
    print('favouritefruit = $favouritefruit');
    print('needspeeling = $needspeeling');
    print('seedsremoved = $seedsremoved');
    print('otherfamily = $otherfamily');
  });
}

File: haskell/rosetta_code/Read_a_configuration_file/Read_a_configuration_file.hs
-----------------------------------------------------------------------------
--
-- Solution for Rosetta Code task "Read a configuration file"
-- <http://rosettacode.org/wiki/Read_a_configuration_file)>
--
-----------------------------------------------------------------------------

import Control.Arrow
import Control.Monad
import Data.Char
import Data.Maybe


main = do
    config <- liftM (getPairs . getLines) $ readFile "fruit.cfg"
    let
        fullname = strVal config "fullname"
        favouritefruit = strVal config "favouritefruit"
        needspeeling = boolVal config "needspeeling"
        seedsremoved = boolVal config "seedsremoved"
        otherfamily = lstVal config "otherfamily"
    putStrLn $ "fullname = " ++ show fullname
    putStrLn $ "favouritefruit = " ++ show favouritefruit
    putStrLn $ "needspeeling = " ++ show needspeeling
    putStrLn $ "seedsremoved = " ++ show seedsremoved
    putStrLn $ "otherfamily = " ++ show otherfamily


boolVal config key = isJust $ lookup key config

strVal config key = maybe "" head $ lookup key config

lstVal config key = fromMaybe [] $ lookup key config

getPairs = map ((parseKey *** parseValue) . getPair)
    where
        parseKey = map toLower
        parseValue = map lstrip . split ','
        getPair = span (not . isSep)

getLines content = filter relevant $ map lstrip $ lines content

lstrip = dropWhile isSpace 

relevant [] = False
relevant (c : _) = not $ elem c ['#', ';']

isSep c = elem c [' ', '=']

split _ [] = []
split sep l@(x : xs)
    | x == sep = split sep xs
    | otherwise = prefix : split sep sufix
        where (prefix, sufix) = span (/= sep) l

